woocommerce/woocommerce
# [Navigation] Prepare to remove feature

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Partially fix for https://github.com/woocommerce/woocommerce/issues/48997

WooCommerce Navigation was introduced a few years ago behind a feature flag (`navigation`) that also required an option to be turned on via CLI (`woocommerce_navigation_enabled`). This PR removes all client side Navigation code and prepares to remove the feature and all its associated classes in PHP. This requires careful consideration because during development we asked the community to register with the new Navigation.

* Public blog post: [Registering with WooCommerce Admin Navigation](https://developer.woocommerce.com/extension-developer-guide/registering-with-woocommerce-admin-navigation/)
* AutomateWoo's resulting changes, as an example: [Register WooCommerce navigation items #836](https://github.com/woocommerce/automatewoo/pull/836)

The recommended way to interact with these classes was to first check if the Navigation classes exist:

```php
if ( class_exists( 'Automattic\WooCommerce\Admin\Features\Navigation\Menu' ) ) {
	Menu::add_plugin_category(
		array(
			'id'         => 'automatewoo',
			'title'      => __( 'AutomateWoo', 'automatewoo' ),
			'capability' => 'manage_woocommerce',
			'url'        => 'automatewoo',
		)
	);
}
```
If plugin developers (internal and external!) did this correctly, we should be able to remove Navigation classes without issue. However, we can't assume that to be the case and a careful approach should be considered. Here steps we can take to minimize problems for merchants.

1. Force the feature off and remove client side code (done in this PR).
2. Deprecate the Navigation classes and add a deprecation notice (done in this PR).
3. Revert all WooCommerce plugins that have hooked into Navigation. [List here](https://github.com/woocommerce/woocommerce-admin/issues/5335)
4. Communicate the change on our developer blog.
5. Remove Navigation classes



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Its best to test this locally so you can switch to this branch to simulate an update.

1. Using `trunk` locally, install AutomateWoo and Subscriptions. These plugins include code that uses Navigation classes. 
2. Turn on Navigation `wp option set woocommerce_navigation_enabled yes`
3. Visit WooCommerce and see the Navigation in action.
4. Switch to this branch. Make sure no errors occur and check the logs to see deprecation notices. Note that I decided to simply error log the deprecation notice instead of using `wc_deprecated_function` which will display an error at the top of every page in admin. Since WooCommerce plugins use Navigation functions. 
5. Lastly, `PageController` was using `Screen::is_woocommerce_page()` in a function `is_embed_page`, which is an important check to determine what kind of page we are on. I removed that logic from `is_embed_page` because I am fairly certain it always evaluates to false when the Navigation feature isn't in use. It should never have been used here, but I think a sanity check and testing around its removal would be a good idea.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

Remove WooCommerce Navigation client side feature and deprecate PHP classes.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
